src/ide/analysis_progress.rs

```diff
@@ -18,205 +19,207 @@ pub enum ProcMacroServerStatus {
+        self.update_receiver.clone()
     }
 
     pub fn on_config_change(&self, config: &Config) {
-        self.state.lock().unwrap().on_config_change(config)
+        self.send(AnalysisEvent::ConfigLoad { enable_proc_macros: config.enable_proc_macros });
     }
 
-    pub fn try_start_analysis(&self) {
-        self.server_tracker.reset_request_tracker();
-        self.state.lock().unwrap().try_start_analysis()
+    pub fn diagnostic_start(&self) {
+        self.send(AnalysisEvent::DiagnosticsTickStart);
     }
 
-    pub fn try_stop_analysis(&self) {
-        let stopped = self.state.lock().unwrap().try_stop_analysis(
-            self.server_tracker.get_did_submit_procmacro_request(),
-            self.server_tracker.get_did_submit_procmacro_request_previously(),
-            self.server_tracker.get_server_status(),
-        );
-        if stopped {
-            self.server_tracker.reset_previous_request_tracker();
-        } else {
-            self.server_tracker.store_previous_request_tracker();
-        }
+    pub fn diagnostic_end(&self, was_cancelled: bool) {
+        self.send(AnalysisEvent::DiagnosticsTickEnd {
+            was_cancelled,
+            request_count: self.server_tracker.get_procmacro_request_count(),
+        });
+    }
+
+    pub fn mutation(&self) {
+        self.send(AnalysisEvent::Mutation);
+    }
+
+    pub fn server_tracker(&self) -> ProcMacroServerTracker {
+        self.server_tracker.clone()
+    }
+
+    fn send(&self, event: AnalysisEvent) {
+        let _ = self.server_tracker.events_sender.send(event);
     }
 }
 
 impl AnalysisProgressController {
-    pub fn new(notifier: Notifier, server_tracker: ProcMacroServerTracker) -> Self {
-        Self {
-            server_tracker,
-            state: Arc::new(Mutex::new(AnalysisProgressControllerState::new(notifier))),
-        }
+    pub fn new(notifier: Notifier) -> Self {
+        let (status_sender, update_receiver) = crossbeam::channel::unbounded();
+        let (events_sender, events_receiver) = crossbeam::channel::unbounded();
+        let server_tracker = ProcMacroServerTracker::new(events_sender);
+        let update_thread = AnalysisProgressThread::spawn(events_receiver, status_sender, notifier);
+
+        Self { server_tracker, update_receiver, _update_thread: Arc::new(update_thread) }
     }
 }
 
 // We don't need to track starts for now
 #[derive(PartialEq)]
-pub enum AnalysisProgressStatus {
-    ResolvedAllProcMacros,
-}
-
-#[derive(Clone)]
-struct AnalysisProgressUpdateChannels {
-    sender: Sender<AnalysisProgressStatus>,
-    receiver: Receiver<AnalysisProgressStatus>,
+pub struct AnalysisFinished;
+
+enum AnalysisEvent {
+    ConfigLoad {
+        /// Loaded asynchronously from config
+        enable_proc_macros: bool,
+    },
+    Mutation,
+    ApplyResponses {
+        response_count: u64,
+    },
+    DiagnosticsTickStart,
+    DiagnosticsTickEnd {
+        was_cancelled: bool,
+        request_count: u64,
```

Comment:
```suggestion
        all_request_count: u64,
```
Add a doc comment explaining what this number is exactly pls